src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/WattTimeDataSourceTests.cs

```diff
@@ -308,15 +252,70 @@ public void GetCarbonIntensity_ConvertsMoerToGramsPerKwh(float lbsPerMwhEmission
         Assert.That(result * GRAMS_TO_POUNDS / KWH_TO_MWH, Is.EqualTo(lbsPerMwhEmissions).Within(FLOATING_POINT_TOLERANCE));
     }
 
-    private void SetupBalancingAuthority(BalancingAuthority balancingAuthority, Location location)
+    /// <summary>
+    /// Tests that if 'frequency' is not provided in the WattTime response of emission data, it is calculated from the first 2 data points, or defaulted to 0 if fewer than 2 data points are returned 
+    /// </summary>
+    [TestCase(new double[] { 300, 300 }, 300, null, TestName = "GetCarbonIntensity - for multiple data points, frequency is null for one data point ")]
+    [TestCase(new double[] { 0 }, null, TestName = "GetCarbonIntensity - for less than 2 data points, frequency is null for one data point ")]
+    [TestCase(new double[] { 300, 300 }, null, null, TestName = "GetCarbonIntensity - for multiple data points, frequency is null for all data points")]
+    [TestCase(new double[] { 500 }, 500, TestName = "GetCarbonIntensity - frequency is not null")]
+    public async Task GetCarbonIntensity_CalculatesDurationBasedOnFrequency(double[] durationValues, params int?[] frequencyValues)
     {
-        this.LocationSource.Setup(r => r.ToGeopositionLocationAsync(location)).Returns(Task.FromResult(location));
-        var latitude = location.Latitude.ToString() ?? throw new ArgumentNullException(String.Format("Could not find location"));
-        var longitude = location.Longitude.ToString() ?? throw new ArgumentNullException(String.Format("Could not find location"));
+        // Arrange
+        var startDate = this.DefaultDataStartTime;
+        var endDate = new DateTimeOffset(2022, 4, 18, 12, 50, 00, TimeSpan.FromHours(-6));
+        var emissionData = GenerateDataPoints(frequencyValues.Length);
+        for( int i = 0; i < frequencyValues.Length; i++)
+        {
+            emissionData[i].Frequency = frequencyValues[i];
+        }
+        
+        List<double> expectedDurationList = durationValues.ToList<double>();
+
+        this.WattTimeClient.Setup(w => w.GetDataAsync(
+            It.IsAny<BalancingAuthority>(),
+            It.IsAny<DateTimeOffset>(),
+            It.IsAny<DateTimeOffset>())
+        ).ReturnsAsync(() => emissionData);
+
+        // Act
+        var result = await this.DataSource.GetCarbonIntensityAsync(new List<Location>() { this.DefaultLocation }, startDate, endDate);
+
+        // Assert
+        List<double> actualDurationList = result.Select(e => e.Duration.TotalSeconds).ToList();
+        
+        CollectionAssert.AreEqual(expectedDurationList, actualDurationList);
```

Comment:
Does `CollectionAssert` work if the elements aren't in the same order? Or do we always expect the test case to match the output order of the durations

Reply:
CollectionAssert checks for the order of elements in the collection too